vignettes/data-preparation.Rmd

```diff
@@ -0,0 +1,399 @@
+dfGroups <- lReporting$Reporting_Groups
+dfMetrics <- lReporting$Reporting_Metrics
+
+# Create dfAnalyticsInput for participant-level drill-down
+dfAnalyticsInput <- purrr::map(lAnalyzed, "Analysis_Input") %>%
+  dplyr::bind_rows(.id = "MetricID") %>%
+  dplyr::as_tibble() %>%
+  dplyr::mutate(SnapshotDate = Sys.Date())
+```
+
+The reporting data model is documented in detail at [Reporting Data](https://gilead-biostats.github.io/gsm.core/articles/DataModel.html#reporting-data).
+
+# App Deployment
+
+Now that our data are ready let's look at how to deploy an app.
+
+## Creating the fnFetchData Function
+
+The `fnFetchData` function is essential for the {gsm.app} to provide drill-down functionality for domain-specific data. This function tells the app how to fetch domain data and should return filtered data based on the parameters passed from the app interface.
+
+```{r create-fnfetchdata, eval = FALSE}
+fnFetchData <- function(strDomainID, strGroupLevel = NULL, strGroupID = NULL, 
+                       strSubjectID = NULL, dSnapshotDate = NULL) {
+  
+  # Use the sample function as a template
+  # In your implementation, replace this with code that fetches from your data sources
+  gsm.app::sample_fnFetchData(
+    strDomainID = strDomainID,
+    strGroupLevel = strGroupLevel, 
+    strGroupID = strGroupID,
+    strSubjectID = strSubjectID,
+    dSnapshotDate = dSnapshotDate
+  )
+}
+```
+
+
+## Running the App
+
+Once you have prepared all the required data structures and the `fnFetchData` function, you can deploy the {gsm.app}.
+
+For development and testing, run the app locally: run_gsm_app() creates a Shiny app to explore a set of clinical trial data. The app facilitates exploration of the data by allowing the user to click to diver deeper into aspects of the data.
+
+
+```{r local-deployment, eval = FALSE}
+# Create the app data structure
+run_gsm_app(
+  dfAnalyticsInput = dfAnalyticsInput,
+  dfBounds = dfBounds,
+  dfGroups = dfGroups,
+  dfMetrics = dfMetrics,
+  dfResults = dfResults,
+  fnFetchData = fnFetchData,
+  fnCountData = ConstructDataCounter(fnFetchData),
+  chrDomains = c(AE = "Adverse Events", DATACHG = "Data Changes", DATAENT = "Data Entry",
+    ENROLL = "Enrollment", LB = "Lab", PD = "Protocol Deviations", QUERY = "Queries",
+    STUDCOMP = "Study Completion", SUBJ = "Subject Metadata", SDRGCOMP =
+    "Treatment Completion"),
+  lPlugins = NULL,
+  strTitle = ExtractAppTitle(dfGroups),
+  strFavicon = "angles-up",
+  strFaviconColor = "#FF5859",
+  tagListExtra = NULL,
+  fnServer = NULL
+)
+```
+
+# Troubleshooting
+
+## Common Issues
+
+1.  **Data Validation Errors**: Ensure all required columns are present and correctly named
+2.  **Missing MetricID Links**: Verify MetricID values are consistent across all data frames
+3.  **Date Format Issues**: Ensure all date columns are properly formatted as Date objects
+4.  **Memory Issues**: For large datasets, consider data filtering and optimization
```

Comment:
Make this sound more specific to the type of data we're looking at ("For extremely large clinical trials" or something along those lines), and offer more specific tips. I'm actually not sure what's valid as far as subdividing the core data. Realistically this is probably only really a potential issue for domain data. We pass errors on from `fnFetchData()` to users, so (here or in the `fnFetchData()` section), for example, I'd recommend having fnFetchData() throw an error suggesting that the user choose a site and/or participant if those values come to fnFetchData as NULL and the data is too big to load.